tests/integrationv2/test_external_psk.py

```diff
@@ -0,0 +1,294 @@
+                      'c388e93343694093934ae4d357'
+
+shared_psk_secret = '4ecd0eb6ec3b4d87f5d6028f922ca4c5851a277fd41311c9e62d2c9492e1c4f3'
+
+s2n_known_value_psk_sha256 = shared_psk_identity + ',' + shared_psk_secret + ',' + 'SHA256'
+
+s2n_known_value_psk_sha384 = shared_psk_identity + ',' + shared_psk_secret + ',' + 'SHA384'
+
+openssl_known_value_psk_parameters = [ '-psk_identity', shared_psk_identity, '--psk', shared_psk_secret ]
+
+
+# Arbitrary Test vectors
+s2n_psk_parameters_sha384 = 'psk_identity_sha384,'\
+                            'a6dadae4567876,'\
+                            'SHA384'\
+
+s2n_psk_parameters_sha256 = 'psk_identity_sha256,'\
+                            'a64dafcd0fc67d2a,'\
+                            'SHA256'\
+
+
+# PSK Identity marker to assert with openssl logs
+psk_identity_marker_openssl_client = [ b"2c0",
+                                       b"35d829359ee5ff7a", b"f4ec900000000262", b"a6494dc486d2c8a3", 
+                                       b"4cb33fa90bf1b007", b"0ad3c498883c9367", b"c09a2be785abc55c", 
+                                       b"d226097a3a982117", b"283f82a03a143efd", b"3ff5dd36d64e861b", 
+                                       b"e7fd61d2827db279", b"cce145077d454a36", b"64d4e6da4d29ee03", 
+                                       b"725a6a4dafcd0fc6", b"7d2aea70529513e3", b"da2677fa5906c5b3",
+                                       b"f7d8f92f228bda40", b"dda721470f9fbf29", b"7b5aea617646fac5", 
+                                       b"c03272e970727c62", b"1a79141ef5f7de65", b"05e5bfbc388e9334", 
+                                       b"3694093934ae4d35", b"7.." ]
+
+
+psk_identity_marker_openssl_server = [ b"2c035d829359",
+                                       b"ee5ff7af4ec90000", b"0000262a6494dc48", b"6d2c8a34cb33fa90",
+                                       b"bf1b0070ad3c4988", b"83c9367c09a2be78", b"5abc55cd226097a3",
+                                       b"a982117283f82a03", b"a143efd3ff5dd36d", b"64e861be7fd61d28",
+                                       b"27db279cce145077", b"d454a3664d4e6da4", b"d29ee03725a6a4da",
+                                       b"fcd0fc67d2aea705", b"29513e3da2677fa5", b"906c5b3f7d8f92f2",
+                                       b"28bda40dda721470", b"f9fbf297b5aea617", b"646fac5c03272e97",
+                                       b"0727c621a79141ef", b"5f7de6505e5bfbc3", b"88e9334369409393",
+                                       b"4ae4d357" ]
+
+
+class Outcome(enum.Enum):
+    psk_connection = auto()
+    full_handshake = auto()
+    connection_failed = auto()
+
+S2N_TEST_CASE = namedtuple(
+    'S2N_TEST_CASE', 'cipher certificate s2nc_params s2nd_params outcome')
+
+S2N_TEST_SUITE = [
+    S2N_TEST_CASE(cipher=Ciphers.CHACHA20_POLY1305_SHA256,
+                  certificate=None,
+                  s2nd_params=[ '--psk', s2n_known_value_psk_sha256 ],
+                  s2nc_params=[ '--psk', s2n_known_value_psk_sha256 ],
```

Comment:
Instead of parameterizing with the exact command params, could you just parameterize with identity, secret, and hash? Then you could use the same test cases for all providers, albeit with maybe different outcomes.

Reply:
I think it will complicate  the testing and make the tests with multiple if-statements to guess the outcome. In the current setup it is a test vector with outcomes. Basically, we have the following for each peer here for PSK parameters: 

- PSK params
    - psk_secret
    - psk_identity 
    - psk_hash_alg
 - Cipher Algorithms 
 - Certificates 
 - Outcome based on psk input and certs 

Only If we match all three params (psk identity, secret, and hash algorithm) for client and server then we should have a successful psk_connection.  The complication comes from we have only limited set of hash algorithms, TLS13 ciphers and curves which we can use for PSK connections. My current test cases cover all of these cases.

Reply:
I tried a different approach and cleaned up the code to be more maintainable, extendable and easy to understand with comments.